correct header guards and add missing headers and declarations

fwph opened the pull request:
# Description

This PR is a superset of the previously opened PR https://github.com/intel/qpl/pull/38

The code changes in this PR were generated automatically by the Permanence AI Coder and reviewed by myself. The goal of this PR is to fix several buggy header files and improve compatibility with C++. The changes in this bugfix PR include:

- updates several header files to add include guards which were previously missing
- update two files to fix include guards which were incorrect.
- adds explicit declarations for implicitly extern functions in .c sources which did not previously have declarations in headers
- add header files as needed to support the new declarations

on-behalf-of: @permanence-ai [github-ai@permanence.ai](mailto:github-ai@permanence.ai)

# Checklist

- No new tests as this is purely a build-time correctness update
- Software path test results had only expected failures (per latest release notes) following the changes:

```
[----------] Global test environment tear-down
[==========] 385 tests from 125 test suites ran. (444798 ms total)
[  PASSED  ] 334 tests.
[  SKIPPED ] 45 tests, listed below:
. . .
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] ta_c_api_deflate_stateful.dynamic_default_verify
[  FAILED  ] ta_c_api_deflate_stateful.dynamic_high_verify
[  FAILED  ] ta_c_api_deflate_stateful.fixed_default_verify
[  FAILED  ] ta_c_api_deflate_stateful.fixed_high_verify
[  FAILED  ] ta_c_api_deflate_stateful.static_default_verify
[  FAILED  ] ta_c_api_deflate_stateful.static_high_verify

 6 FAILED TESTS
  YOU HAVE 22 DISABLED TESTS
```

mzhukova commented on sources/c_api/legacy_hw_path/qpl_hw_get_size.h:
```diff
@@ -0,0 +1,15 @@
+/*******************************************************************************
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ ******************************************************************************/
+
+#ifndef QPL_HW_GET_SIZE_H
+#define QPL_HW_GET_SIZE_H
+
+// includes extracted from sources/c_api/legacy_hw_path/qpl_hw_get_size.c
+#include "own_defs.h"
+#include "hardware_state.h"
+
+
+#endif
```
Unclear on the value of such changes if we're just moving out a couple of header files, so I'd rather avoid it.

fwph replied:
got it! how do you feel about limiting the additional header files to those which are necessary to cover the new declarations?

mzhukova replied:
Do you mean similar to changes in `sources/c_api/compression_operations/qpl_deflate_slow.h`?

fwph replied:
yes, that's one example -- there are a few more in the third commit that are similar.

Permanence-AI-Coder committed:
add missing or correct errors in header guards for 9 files

    - modify sources/core-sw/src/compression/opt/placeholder.h (add missing include guard)
    - modify sources/core-sw/src/filtering/opt/qplc_unpack_16u_k0.h (add missing include guard)
    - modify sources/core-sw/src/filtering/opt/qplc_unpack_8u_k0.h (add missing include guard)
    - modify sources/core-sw/src/filtering/opt/qplc_unpack_be_16u_k0.h (add missing include guard)
    - modify sources/core-sw/src/filtering/opt/qplc_unpack_be_32u_k0.h (add missing include guard)
    - modify sources/core-sw/src/filtering/opt/qplc_unpack_be_8u_k0.h (add missing include guard)
    - modify sources/isal/igzip/huffman.h (add missing include guard)
    - modify sources/isal/igzip/igzip_wrapper.h (add missing #define inside include guard)
    - modify sources/isal/igzip/static_inflate.h (incorrectly located #endif of include guard)

on-behalf-of: @permanence-ai <github-ai@permanence.ai>

Permanence-AI-Coder committed:
add missing extern function declarations to 15 files

    - modify sources/core-sw/src/compression/include/deflate_slow_matcher.h (add missing declarations for ChangeRequest(file=PosixPath('/home/permai/workspace/qpl-working/sources/core-sw/src/compression/include/deflate_slow_matcher.h'), line=-1, type='decl', symbol=func get_lazy_best_match @ sources/core-sw/src/compression/deflate_slow_matcher.c:125, change_type=<ChangeType.ADD_DECLARATION: 3>, change_value=None))
    - create sources/isal/igzip/adler32_base.h (add missing header file)
    - modify sources/isal/igzip/adler32_base.c (add #include of new header file)
    - modify sources/isal/igzip/encode_df.h (add missing declarations for ChangeRequest(file=PosixPath('/home/permai/workspace/qpl-working/sources/isal/igzip/encode_df.h'), line=-1, type='decl', symbol=func qpl_encode_deflate_icf_base @ sources/isal/igzip/encode_df.c:24, change_type=<ChangeType.ADD_DECLARATION: 3>, change_value=None))
    - modify sources/isal/igzip/huff_codes.h (add missing declarations for ChangeRequest(file=PosixPath('/home/permai/workspace/qpl-working/sources/isal/include/igzip_lib.h'), line=621, type='decl', symbol=decl qpl_isal_create_hufftables @ sources/isal/include/igzip_lib.h:621, change_type=<ChangeType.ADD_EXTERN_C: 1>, change_value=None) ChangeRequest(file=PosixPath('/home/permai/workspace/qpl-working/sources/isal/include/igzip_lib.h'), line=634, type='decl', symbol=decl qpl_isal_create_hufftables_subset @ sources/isal/include/igzip_lib.h:634, change_type=<ChangeType.ADD_EXTERN_C: 1>, change_value=None) ChangeRequest(file=PosixPath('/home/permai/workspace/qpl-working/sources/isal/igzip/huff_codes.h'), line=151, type='decl', symbol=decl qpl_create_hufftables_icf @ sources/isal/igzip/huff_codes.h:151, change_type=<ChangeType.ADD_EXTERN_C: 1>, change_value=None) ChangeRequest(file=PosixPath('/home/permai/workspace/qpl-working/sources/isal/igzip/huff_codes.h'), line=-1, type='decl', symbol=func qpl_isal_update_histogram_base @ sources/isal/igzip/huff_codes.c:643, change_type=<ChangeType.ADD_DECLARATION: 3>, change_value=None))
    - create sources/isal/igzip/igzip.h (add missing header file)
    - modify sources/isal/igzip/igzip.c (add #include of new header file)
    - create sources/isal/igzip/igzip_base.h (add missing header file)
    - modify sources/isal/igzip/igzip_base.c (add #include of new header file)
    - create sources/isal/igzip/igzip_icf_base.h (add missing header file)
    - modify sources/isal/igzip/igzip_icf_base.c (add #include of new header file)
    - create sources/isal/igzip/igzip_icf_body.h (add missing header file)
    - modify sources/isal/igzip/igzip_icf_body.c (add #include of new header file)
    - create sources/isal/igzip/igzip_inflate.h (add missing header file)
    - modify sources/isal/igzip/igzip_inflate.c (add #include of new header file)

on-behalf-of: @permanence-ai <github-ai@permanence.ai>

fwph commented:
@mzhukova This branch has been regenerated with adjustments to the header guard style, and reducing the number of headers created just to those which are needed for additional declarations. Note the limit here is that it does not attempt to create new extern declarations for functions that use a macro to define the signature.

mzhukova commented on sources/core-sw/src/compression/include/deflate_slow_matcher.h:
```diff
@@ -24,4 +24,11 @@ deflate_match_t get_lazy_best_match(const deflate_hash_table_t *const hash_table
 }
 #endif
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+    deflate_match_t get_lazy_best_match(const deflate_hash_table_t * const hash_table_ptr, const uint8_t * const lower_bound_ptr, const uint8_t * const string_ptr, const uint8_t * const upper_bound_ptr);
+#ifdef __cplusplus
+} // extern "C"
+#endif
```
I've looked into changes more carefully and actually, this seems to be duplication of L18-L22 of the same file.

mzhukova commented on sources/isal/igzip/adler32_base.h:
```diff
@@ -0,0 +1,25 @@
+ * SPDX-License-Identifier: MIT
+ ******************************************************************************/
+
+#ifndef QPL_SOURCES_ISAL_IGZIP_ADLER32_BASE_C
+#define QPL_SOURCES_ISAL_IGZIP_ADLER32_BASE_C
+
+// includes extracted from sources/isal/igzip/adler32_base.c
+#include <stddef.h>
+#include <stdint.h>
+#include "igzip_checksums.h"
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+    uint32_t qpl_adler32_base(uint32_t adler32, uint8_t *start, uint32_t length);
+#ifdef __cplusplus
+} // extern "C"
+#endif
```
Unclear on this change goal. Is this header supposed to be included into `sources/isal/igzip/adler32_base.c` or to the `.c` or `.cpp` files where we're calling qpl_adler32_base (so that the declaration is readily available)?

For instance, this particular `qpl_adler32_base` function is used in a single `.asm` file that relies on extern, so probably not necessary to make its declaration available.
And it is also not required for `sources/isal/igzip/adler32_base.c` where it is being included currently.

Additionally, what is the reason for having these here -- 
...
#include <stddef.h>
#include <stdint.h>
#include "igzip_checksums.h"
... 
They are not required for using the `qpl_adler32_base` (this would probably require just including the header for uint32_t). And they are already present in `sources/isal/igzip/adler32_base.c`.

mzhukova replied:
++ similar comment for changes below.

In fact, these `foo_base` functions are an interesting case. They are supposed to be a reference implementation, and dispatching to either reference or optimized version is handled in some `.asm` as I mentioned earlier. So, really the `foo_base` in most of the cases should never be used directly, therefore no need for such declarations, the rest of the library would be using `foo` that is, in some sense, a wrapper that would dispatch to available optimization.

jweese replied:
The intent here (and similar below) is to be able to compile `adler32_base.c` with either a c or c++ compiler. If we use a c++ compiler, since there is no explicit declaration of language linkage, the symbol for `qpl_adler32_base` gets name-mangled, so we get a linker error since, as you pointed out, the callsite in the `asm` file sees it `extern` and expects the umangled symbol.

With this change included into the `.c` file that has the function definition, we're sure to always get C language linkage and an unmangled symbol.

You are right that, since the `asm` file never sees this declaration, we could fix the linkage inside the `.c` file instead of introducing this new header, if that is what you prefer.

mzhukova commented on sources/isal/igzip/igzip.h:
```diff
@@ -0,0 +1,52 @@
+ * SPDX-License-Identifier: MIT
+ ******************************************************************************/
+
+#ifndef QPL_SOURCES_ISAL_IGZIP_IGZIP_C
```
I guess, this should be _H

mzhukova commented on sources/isal/igzip/igzip.h:
```diff
@@ -0,0 +1,52 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+    void qpl_isal_deflate_body_huffman_only(struct isal_zstream *stream);
+    void qpl_isal_deflate_finish_huffman_only(struct isal_zstream *stream);
+    void qpl_isal_deflate_hash(struct isal_zstream *stream, uint8_t *dict, uint32_t dict_len);
```
Could we instead include these declarations into `igzip.c`? (also again, this doesn't seem necessary right now).

mzhukova commented on sources/isal/igzip/igzip.h:
```diff
@@ -0,0 +1,52 @@
+#define QPL_SOURCES_ISAL_IGZIP_IGZIP_C
+
+// includes extracted from sources/isal/igzip/igzip.c
+#include <assert.h>
+#include <string.h>
+#ifdef _WIN32
+# include <intrin.h>
```
These are still present in `igzip.c`, so the change doesn't really make sense to me. Also, not all declarations/externs were moved to the header, so really I'd just prefer 3 declarations to be inserted into the `igzip.c` directly.

mzhukova commented:
I think this one require rework and rebase, so will be moving to draft state meanwhile.
